src/frequenz/repo/config/version.py

```diff
@@ -0,0 +1,443 @@
+expected follow the format:
+
+- `vX.x.x` for major branches, where `X` is the major version number. For example,
+  `v1.x.x` is the major branch for the major version 1.
+
+  It represents an in-development major version, from which new minor branches
+  including new features are created (including the first minor branch for the major
+  version).
+
+- `vX.Y.x` for minor branches, where `X` is the major version number and `Y` is the
+  minor version number. For example, `v1.0.x` is the minor branch for the major version
+  1 and minor version 0.
+
+  It represents a maintained minor version, from which new patch releases (with bug
+  fixes) are created. For example, `v1.1.x` is the minor branch for the major version 1
+  and minor version 1.
+
+[semver]: https://semver.org/
+"""
+
+from __future__ import annotations
+
+import dataclasses
+import logging
+import pathlib
+import re
+from typing import Self, TypeGuard
+
+import semver
+
+_logger = logging.getLogger(__name__)
+_major_branch_re: re.Pattern[str] = re.compile(r"^v?(\d+)\.x\.x$")
+_minor_branch_re: re.Pattern[str] = re.compile(r"^v?(\d+)\.(\d+)\.x$")
+
+
+def strip_v(version: str) -> str:
+    """Strip the leading 'v' from a version string.
+
+    Args:
+        version: The version string.
+
+    Returns:
+        The version string without the leading 'v'.
+    """
+    return version[1:] if version.startswith("v") else version
+
+
+def to_semver(version: str) -> semver.Version | None:
+    """Convert a version string to a semantic version.
+
+    Args:
+        version: The version string.
+
+    Returns:
+        The semantic version or `None` if the version string is not a valid semantic
+            version.
+    """
+    try:
+        return semver.Version.parse(strip_v(version))
+    except ValueError:
+        logging.debug("Version is not a semantic version: %s", version)
+        return None
+
+
+def _build_tags(tags: list[str]) -> dict[str, semver.Version]:
+    """Build the tags dictionary.
+
+    Args:
+        tags: The tags of the repository.
+
+    Returns:
+        The tags dictionary, where the key is the tag name and the value is the
+            parsed tag (semantic) version, sorted by semantic version.
+    """
+
+    def tag_not_none(
+        tag: tuple[str, semver.Version | None]
+    ) -> TypeGuard[tuple[str, semver.Version]]:
+        return tag[1] is not None
+
+    return dict(
+        sorted(
+            filter(
+                tag_not_none,
+                ((tag, to_semver(tag)) for tag in (tags or [])),
+            ),
+            key=lambda tag: tag[1],
+        )
+    )
+
+
+def _build_branches(branches: list[str]) -> dict[str, BranchVersion]:
+    """Build the branches dictionary.
+
+    Args:
+        branches: The branches of the repository.
+
+    Returns:
+        The branches dictionary, where the key is the branch name and the value is
+            the parsed branch name, sorted by branch version.
```

Comment:
I am slightly confused by the `sorted` part, a `dict` isn't sorted (I thought) and otherwise and the value is a `BranchVersion` which seems to be a single value, not something that contains ordered values?

Edit: Okay, I read up a bit and it seems dict's are actually ordered in the sense that they keep their insertion order

Reply:
Yeah, `dict` preserve order of insertion when iterating. I'm not sure what to do here, for once it is a slightly obscure feature that not everybody knows about, but then it is just regular Python. But I guess a comment wouldn't hurt either.

Reply:
Mmm, looking at the code I'm not so sure about the clarification. I think adding how Python works to the documentation is too much and if I make it just a code comment then it is not adding a lot of extra clarity, especially if we have to assume the end user (only reading the docs) already should know about dicts preserving order.

Reply:
Oh, wait, I know see it is actually missing the sorting for the branch function (the tag has it), I guess it was lost in some rebase because I'm pretty sure I added it to both at some point.